foundations/foundation-metrics/src/main/java/io/servicecomb/foundation/metrics/performance/MetricsDataMonitorUtil.java

```diff
@@ -0,0 +1,236 @@
+
+    // store average, min and max times in instance level.
+    QueueMetrics reqMetricsInstance = MetricsDataMonitor.getMetricsGlobalRef()
+        .getQueueMetrics("instance_level_key");
+    if (reqMetricsInstance == null) {
+      reqMetricsInstance = new QueueMetrics();
+      reqMetricsInstance.setMinLifeTimeInQueue(timeInqueue);
+      reqMetricsInstance.setMaxLifeTimeInQueue(timeInqueue);
+      reqMetricsInstance.setAvgLifeTimeInQueue(avgTimeInQueueInstance);
+      reqMetricsInstance.setCountInQueue(MetricsDataMonitor.getMetricsGlobalRef().getQueueCount());
+      MetricsDataMonitor.getMetricsGlobalRef().setQueueMetrics("instance_level_key", reqMetricsInstance);
+    } else {
+      reqMetricsInstance.setCountInQueue(metricsRef.getQueueCount());
+      reqMetricsInstance.setAvgLifeTimeInQueue(avgTimeInQueueInstance);
+
+      long mintime = reqMetricsInstance.getMinLifeTimeInQueue().get();
+      if (mintime > timeInqueue) {
+        reqMetricsInstance.setMinLifeTimeInQueue(timeInqueue);
+      }
+
+      long maxtime = reqMetricsInstance.getMaxLifeTimeInQueue().get();
+      if (maxtime < timeInqueue) {
+        reqMetricsInstance.setMaxLifeTimeInQueue(timeInqueue);
+      }
+    }
+  }
+
+  /**
+   * Prepare the metrics related to working time after queue.
+   */
+  public static void prepareMetricsAfterQueue(Invocation invocation) {
+
+    if ((invocation == null) || (invocation.getOperationMeta() == null)) {
+      LOGGER.error("Invocation is not correct");
+      return;
+    }
+
+    String operPath = invocation.getOperationMeta().getMicroserviceQualifiedName();
+    QueueMetricTimes queueMetricTimes = invocation.getQueueMetricTimes();
+    if (queueMetricTimes == null) {
+      LOGGER.error("Queue time metrics can not be NULL");
+      return;
+    }
+
+    invocation.getQueueMetricTimes().setEndOpertime(System.currentTimeMillis());
+
+    long serviceExecTime = queueMetricTimes.getEndOpertime() - queueMetricTimes.getQueueEndtime();
+    QueueMetricsData reqQueue = MetricsDataMonitor.getMetricsGlobalRef().getOrCreateOperationMetrics(operPath);
+    reqQueue.getTotalServExecutionTime().addAndGet(serviceExecTime);
+    reqQueue.incrementTotalServExecutionCount();
+
+    MetricsDataMonitor metricRef = MetricsDataMonitor.getMetricsGlobalRef();
+    metricRef.getTotalServExecutionTime().addAndGet(serviceExecTime);
+    metricRef.incrementTotalServExecutionCount();
+
+    double avgWorkTimeAfterQueue = reqQueue.getTotalServExecutionTime().get()
+        / reqQueue.getTotalServExecutionCount().get();
+
+    QueueMetrics requestMetricQueue = MetricsDataMonitor.getMetricsGlobalRef().getQueueMetrics(operPath);
+    if (requestMetricQueue == null) {
+      LOGGER.error("Request Metrics Queue can not be NULL");
+      return;
+    }
+    requestMetricQueue.setAvgLifeTimeAfterQueue(avgWorkTimeAfterQueue);
+
+    double avgServiceExecTimeInstance = metricRef.getTotalServExecutionTime().get()
```

Comment:
long must convert to double before division , please check all submit

Reply:
fixed